core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/jupyter/KotlinNotebookPluginUtils.kt

```diff
@@ -79,27 +88,80 @@ public object KotlinNotebookPluginUtils {
+     * @return The sorted dataframe.
      */
-    public fun sortByColumns(df: AnyFrame, columnPaths: List<List<String>>, isDesc: List<Boolean>): AnyFrame =
-        df.sortBy {
-            require(columnPaths.all { it.isNotEmpty() })
-            require(columnPaths.size == isDesc.size)
+    public fun sortByColumns(df: AnyFrame, columnPaths: List<List<String>>, isDesc: List<Boolean>): AnyFrame {
+        require(columnPaths.all { it.isNotEmpty() })
+        require(columnPaths.size == isDesc.size)
+
+        val sortKeys = columnPaths.map { path ->
+            ColumnPath(path)
+        }
+
+        val comparator = createComparator(sortKeys, isDesc)
 
-            val sortKeys = columnPaths.map { path ->
-                ColumnPath(path)
+        return df.sortWith(comparator)
+    }
+
+    private fun createComparator(sortKeys: List<ColumnPath>, isDesc: List<Boolean>): Comparator<DataRow<*>> {
+        return Comparator { row1, row2 ->
+            for ((key, desc) in sortKeys.zip(isDesc)) {
+                val comparisonResult = if (row1.df().getColumn(key).type.isComparable()) {
+                    compareComparableValues(row1, row2, key, desc)
+                } else {
+                    compareStringValues(row1, row2, key, desc)
+                }
+                // If a comparison result is non-zero, we have resolved the ordering
+                if (comparisonResult != 0) return@Comparator comparisonResult
             }
+            // All comparisons are equal
+            0
+        }
+    }
 
-            (sortKeys zip isDesc).map { (key, desc) ->
-                if (desc) key.desc() else key
-            }.toColumnSet()
+    @Suppress("UNCHECKED_CAST")
+    private fun compareComparableValues(
+        row1: DataRow<*>,
+        row2: DataRow<*>,
+        key: ColumnPath,
+        desc: Boolean,
+    ): Int {
+        val firstValue = row1.getValueOrNull(key) as Comparable<Any?>?
+        val secondValue = row2.getValueOrNull(key) as Comparable<Any?>?
+
+        return when {
+            firstValue == null && secondValue == null -> 0
+            firstValue == null -> if (desc) 1 else -1
+            secondValue == null -> if (desc) -1 else 1
+            desc -> secondValue.compareTo(firstValue)
+            else -> firstValue.compareTo(secondValue)
         }
+    }
+
+    private fun compareStringValues(
+        row1: DataRow<*>,
+        row2: DataRow<*>,
+        key: ColumnPath,
+        desc: Boolean,
+    ): Int {
+        val firstValue = (row1.getValueOrNull(key)?.toString() ?: "")
+        val secondValue = (row2.getValueOrNull(key)?.toString() ?: "")
+
+        return if (desc) {
+            secondValue.compareTo(firstValue)
+        } else {
+            firstValue.compareTo(secondValue)
+        }
+    }
+
+    private fun KType.isComparable(): Boolean = this.isSubtypeOf(typeOf<Comparable<*>>())
```

Comment:
This check breaks whenever you have, say, String and Int in a single column. Yes, they are both `Comparable`, but not comparable to each other. They are `Comparable<Nothing>` and calling the `compare` there will break.

Check out the fix I made here for `AnyCol.isComparable()`: https://github.com/Kotlin/dataframe/pull/726/files#diff-68de50cc685a57fef49b535d54b8dc925caecc39ba0bc1de3d312ffeb25e8a60

In this case, if we get `Comparable<Nothing>` we should probably just compare them as `String` again, so returning `false` here should be fine (I think).